Integrations/python/test/test_learn_gather.py

```diff
@@ -0,0 +1,200 @@
+        )).T
+        cls.int_array = np.vstack((
+            np.arange(0, 100, dtype = np.intc),
+            np.arange(100, 0, -1, dtype = np.intc),
+            np.arange(-101, -1, dtype = np.intc)
+        )).T
+        cls.long_array = np.vstack((
+            np.arange(0, 100, dtype = np.int_),
+            np.arange(100, 0, -1, dtype = np.int_),
+            np.arange(-101, -1, dtype = np.int_)
+        )).T
+        cls.float_array = np.vstack((
+            np.arange(0, 100, dtype = np.single),
+            np.sqrt(np.arange(0, 100, dtype = np.single)),
+            np.sqrt(np.sqrt(np.arange(0, 100, dtype = np.single)))
+        )).T
+        cls.double_array = np.vstack((
+            np.arange(0, 100, dtype = np.double),
+            np.sqrt(np.arange(0, 100, dtype = np.double)),
+            np.sqrt(np.sqrt(np.arange(0, 100, dtype = np.double)))
+        )).T
+
+    # Model for learn to use when dtype = [np.bool_]
+    def boolean_model(self, features):
+        return np.count_nonzero(features, axis = 1) < 2
+
+    # Model for learn to use when dtype = [np.byte, np.short, np.intc, np.int_]
+    def whole_model(self, features):
+        return np.sum(features, axis = 1)
+
+    # Model for learn to use when dtype = [np.single, np.double]
+    def decimal_model(self, features):
+        return np.prod(features, axis = 1)
+
+    # Test boolean data types
+    def test_boolean(self):
+        # self.base_test(np_dtype = np.bool_)
+        return
+
+    # Test byte data types
+    def test_byte(self):
+        self.base_test(np_dtype = np.byte)
+
+    # Test short data types
+    def test_short(self):
+        self.base_test(np_dtype = np.short)
+
+    # Test int data types
+    def test_int(self):
+        self.base_test(np_dtype = np.intc)
+
+    # Test long data types
+    def test_long(self):
+        self.base_test(np_dtype = np.int_)
+
+    # Test float data types
+    def test_float(self):
+        self.base_test(np_dtype = np.single)
+
+    # Test double data types
+    def test_double(self):
+        self.base_test(np_dtype = np.double)
+
+    # The base test, which other tests will be built from
+    def base_test(self, np_dtype):
+        if np_dtype == np.bool_:
+            source = self.bool_table
+            model = self.boolean_model
+        elif np_dtype == np.byte:
+            source = self.byte_table
+            model = self.whole_model
+        elif np_dtype == np.short:
+            source = self.short_table
+            model = self.whole_model
+        elif np_dtype == np.intc:
+            source = self.int_table
+            model = self.whole_model
+        elif np_dtype == np.int_:
+            source = self.long_table
+            model = self.whole_model
+        elif np_dtype == np.single:
+            source = self.float_table
+            model = self.decimal_model
+        elif np_dtype == np.double:
+            source = self.double_table
+            model = self.decimal_model
```

Comment:
This if-block is funky.  It seems like the code would be much more simple if this function just accepted the source and model as input.  By making that small change, you can ditch 20 lines of convoluted code.  e.g. `test_double` knows that the type is a double, but you then call a function that re-determines that the type is a double.